plugins/modules/packaging/language/cpanm.py

```diff
@@ -109,105 +113,75 @@
 
 import os
 
-from ansible.module_utils.basic import AnsibleModule
-
-
-def _is_package_installed(module, name, locallib, cpanm, version):
-    cmd = ""
-    if locallib:
-        os.environ["PERL5LIB"] = "%s/lib/perl5" % locallib
-    cmd = "%s perl -e ' use %s" % (cmd, name)
-    if version:
-        cmd = "%s %s;'" % (cmd, version)
-    else:
-        cmd = "%s;'" % cmd
-    res, stdout, stderr = module.run_command(cmd, check_rc=False)
-    return res == 0
-
-
-def _build_cmd_line(name, from_path, notest, locallib, mirror, mirror_only, installdeps, cpanm, use_sudo):
-    # this code should use "%s" like everything else and just return early but not fixing all of it now.
-    # don't copy stuff like this
-    if from_path:
-        cmd = cpanm + " " + from_path
-    else:
-        cmd = cpanm + " " + name
-
-    if notest is True:
-        cmd = cmd + " -n"
-
-    if locallib is not None:
-        cmd = cmd + " -l " + locallib
-
-    if mirror is not None:
-        cmd = cmd + " --mirror " + mirror
-
-    if mirror_only is True:
-        cmd = cmd + " --mirror-only"
-
-    if installdeps is True:
-        cmd = cmd + " --installdeps"
-
-    if use_sudo is True:
-        cmd = cmd + " --sudo"
-
-    return cmd
-
-
-def _get_cpanm_path(module):
-    if module.params['executable']:
-        result = module.params['executable']
-    else:
-        result = module.get_bin_path('cpanm', True)
-    return result
-
-
-def main():
-    arg_spec = dict(
-        name=dict(default=None, required=False, aliases=['pkg']),
-        from_path=dict(default=None, required=False, type='path'),
-        notest=dict(default=False, type='bool'),
-        locallib=dict(default=None, required=False, type='path'),
-        mirror=dict(default=None, required=False),
-        mirror_only=dict(default=False, type='bool'),
-        installdeps=dict(default=False, type='bool'),
-        system_lib=dict(default=False, type='bool', aliases=['use_sudo']),
-        version=dict(default=None, required=False),
-        executable=dict(required=False, type='path'),
+from ansible_collections.community.general.plugins.module_utils.module_helper import (
+    ModuleHelper, CmdMixin, ArgFormat, ModuleHelperException
+)
+
+
+class CPANMinus(CmdMixin, ModuleHelper):
+    module = dict(
+        argument_spec=dict(
+            name=dict(type='str', aliases=['pkg', 'from_url']),
+            version=dict(type='str'),
+            from_path=dict(type='path', aliases=['from_dir']),
+            notest=dict(default=False, type='bool'),
+            locallib=dict(type='path'),
+            mirror=dict(type='str'),
+            mirror_only=dict(default=False, type='bool'),
+            installdeps=dict(type='bool', default=False),
+            system_lib=dict(type='bool', default=False, aliases=['use_sudo']),
+            executable=dict(type='path'),
+        ),
+        required_one_of=[('name', 'from_path')],
+        mutually_exclusive=[('name', 'from_path'), ('from_path', 'version')],
```

Comment:
This is a breaking change. The old code accepted specifying both (even though it should not have), so this should be deprecated (and if both are specified, the old behavior should be used).

Reply:
@russoz what about this one?

Reply:
It's been a tough week, I am lagging behind. :-) well, I will back down on this one for the sake moving the PR forward, but it's likely I will raise another one later specifically for that.